### PR TITLE
feat: Add proposal for promote-dependencies-automatically

### PR DESCRIPTION
**Pull request by zhy76**
**What type of PR is this?**
/kind feature
/kind documentation

**What this PR does / why we need it**:
Add proposal for promote-dependencies-automatically

**Which issue(s) this PR fixes**:
Part of #4036 

**Special notes for your reviewer**:
none
**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [inline docs/proposals/promote-dependencies-automatically/README.md:18]: I think describe like following better for the paragraph:
One resource could only be okay when the dependencies are ready, such as Deployment depending on configmap. So, If we want to promote the resource and deploy it into multiple clusters, we need to promote the resource and all the dependencies, it would be great if this process could be done automatically.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [inline docs/proposals/promote-dependencies-automatically/README.md:99]: I think it's necessary to indicate that: 
We will implement the webhook requests based on the logic of `port-forward`, but not the native port-forward implementation.

our: data in program->create requests->send requests through the created port-forward connection->webhook pod.
native: data in program->create requests->send requests to port-forward listening address->send requests through the created port-forward connection->webhook pod.

**Comment by jwcesign** [comment]
Hi, @zhy76 Please add the behavior of meeting promoting failure(validation failed or other), and if part dependencies already exist, what should Karmada do?

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [reply docs/proposals/promote-dependencies-automatically/README.md:113]: +1

**Comment by zhy76** [comment]
Thanks, I will refer to your suggestions~

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [inline docs/proposals/promote-dependencies-automatically/README.md:170]: I think this chapter should be moved to the front. Alternatives means the other implementation way

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [inline docs/proposals/promote-dependencies-automatically/README.md:100]: I think we should skip the resource that already exists, instead of abortion.

What do you think? @chaunceyjiang @XiShanYongYe-Chang

**Review by jwcesign — COMMENTED** [review]
Other lgtm

**Comment by jwcesign** [comment]
Any more suggestions about port-forward? @chaunceyjiang

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (zhy76) [reply docs/proposals/promote-dependencies-automatically/README.md:170]: OK

**Comment by jwcesign** [comment]
/cc @chaunceyjiang 

Do you have any more suggestions?

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (chaunceyjiang) [inline docs/proposals/promote-dependencies-automatically/README.md:148]: Do we have similar information output like the following? To prompt users about which resources have been `promoted`.

```
# kubectl create -f test.yaml
deployment.apps/myapp created   
configmap/my-config created
```
for example:

```
`karmadactl promote workload myapp -c member1 -d=true` 
workload.apps/myapp promoted   
configmap/my-config promoted
```


other LGTM

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (chaunceyjiang) [reply docs/proposals/promote-dependencies-automatically/README.md:100]: +1

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (zhy76) [reply docs/proposals/promote-dependencies-automatically/README.md:148]: Yes, like this:
image

**Comment by zhy76** [comment]
Thanks for your suggestions, I will modidy.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (chaunceyjiang) [reply docs/proposals/promote-dependencies-automatically/README.md:148]: Fantastic!

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (XiShanYongYe-Chang) [inline docs/proposals/promote-dependencies-automatically/README.md:27]: What's the difference between the different colored dependency markers in this diagram, I see them marked in three different colors.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (XiShanYongYe-Chang) [inline docs/proposals/promote-dependencies-automatically/README.md:94]: > we need to perform port forwarding to invoke the service within the cluster.

Is it possible to access it outside the cluster?

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (XiShanYongYe-Chang) [reply docs/proposals/promote-dependencies-automatically/README.md:100]: Perhaps interrupting or skipping could be left to the user's choice.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (XiShanYongYe-Chang) [inline docs/proposals/promote-dependencies-automatically/README.md:148]: What is the meaning of the -d parameter, can you elaborate?

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (XiShanYongYe-Chang) [inline docs/proposals/promote-dependencies-automatically/README.md:150]: How is this operation controlled, we should not be able to do this with our current operation.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [reply docs/proposals/promote-dependencies-automatically/README.md:68]: input of command, default to be true

Comment (XiShanYongYe-Chang) [reply docs/proposals/promote-dependencies-automatically/README.md:68]: thx, I get it, and I feel the need to clarify in the proposal.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (zhy76) [reply docs/proposals/promote-dependencies-automatically/README.md:27]: I borrowed this from the diagram here 😸 : https://github.com/karmada-io/karmada/blob/master/docs/proposals/dependencies-automatically-propagation/dependenciesOfDeployment.png
Maybe I need to make it the same color？

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (zhy76) [reply docs/proposals/promote-dependencies-automatically/README.md:68]: Perhaps it would be better to default to `false` for `Deps`? In this way, if the user needs to open the automatic propagation dependency mechanism, then open it.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (XiShanYongYe-Chang) [reply docs/proposals/promote-dependencies-automatically/README.md:27]: Or maybe being able to explain what the different colors mean would be a little better, ignore it for now.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (RainbowMango) [inline docs/proposals/promote-dependencies-automatically/README.md:70]: I think the `karmadactl promote` doesn't take responsibility to create the `ResourceBinding`, right? It just focuses on promoting the dependents of `resource-templates`.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (RainbowMango) [reply docs/proposals/promote-dependencies-automatically/README.md:100]: > I think we should skip the resource that already exists, instead of aborting.

I tend to abort because in case resources already exist, we can't ensure the content of the resource(like Secret/ConfigMap) is exactly the same as in the member cluster.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (RainbowMango) [inline docs/proposals/promote-dependencies-automatically/README.md:168]: Do we need to use `port forward` to request webhook? @jwcesign 
If yes, this chapter should not be presented in `Alternatives`.

Comment (jwcesign) [reply docs/proposals/promote-dependencies-automatically/README.md:168]: We still need it. As the comments: https://github.com/karmada-io/karmada/pull/4100#discussion_r1361744410

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (zhy76) [reply docs/proposals/promote-dependencies-automatically/README.md:70]: Yes.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (zhy76) [reply docs/proposals/promote-dependencies-automatically/README.md:150]: Let me clarify in my proposal.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [reply docs/proposals/promote-dependencies-automatically/README.md:97]: I think it's possible, and easy to implement. Just request the target url.

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (jwcesign) [inline docs/proposals/promote-dependencies-automatically/README.md:73]: Don't need to create PP/CPP for the dependencies, just enable `propagateDeps` in the original resources' policy

**Comment by jwcesign** [comment]
/lgtm

cc @RainbowMango for checking

File: docs/proposals/promote-dependencies-automatically/README.md
Comment (RainbowMango) [inline docs/proposals/promote-dependencies-automatically/README.md:11]: LGTM.
But I think you can put @jwcesign on to the approvers.
```suggestion
- "jwcesign"
```

**Review by RainbowMango — APPROVED** [review]
/approve
/hold
for setting the approvers and leave final decisions to @jwcesign

**Comment by zhy76** [comment]
> /approve /批准 /hold /抓住 for setting the approvers and leave final decisions to 用于设置批准者并将最终决定权留给@jwcesign

added, thanks~

**Comment by jwcesign** [comment]
/approve

Thank you for your excellent work, @zhy76

**Comment by RainbowMango** [comment]
> Pending — Not mergeable. Job e2e test has not succeeded.

I guess the robot missed the e2e test signal. I'll retrigger the tests again.

**Comment by RainbowMango** [comment]
> > Pending — Not mergeable. Job e2e test has not succeeded.
> 
> I guess the robot missed the e2e test signal. I'll retrigger the tests again.

I'm not sure why the robot hanging there. 
Just rebased this PR and try another time.

**Comment by zhy76** [comment]
> > > Pending — Not mergeable. Job e2e test has not succeeded.
> > 
> > 
> > I guess the robot missed the e2e test signal. I'll retrigger the tests again.
> 
> I'm not sure why the robot hanging there. Just rebased this PR and try another time.

Thanks~

**Pull request merged**
